### ISSUE TITLE
Add libSQL support to SQLite generator

### ISSUE THREAD

**Issue by plunkettscott** [stephenafamo/bob#304]
In v0.29.0 you removed support for Atlas, which was allowing us to use Bob with [libSQL/Turso](https://github.com/tursodatabase/libsql-client-go) (a database not directly supported by Bob). Since Bob is now focused on each database implementation, rather than schema tooling, supporting libSQL would allow us to use the generator directly on our database. To get around the removal in v0.29.0 we have to export our schema from libSQL to a local file to be read by Bob, which works, but adds unnecessary steps to generating code.

The libSQL project is a fork of SQLite, so nothing aside from the driver should need to change to support libSQL. The dialects are identical from a schema perspective.

**Comment by stephenafamo** [comment]
I'll look into this

**Comment by plunkettscott** [comment]
Thanks @stephenafamo !

I'm willing to work on this as well if you give me some direction as to how you'd want this configurable. The way most projects do this is by supporting various connection strings for SQLite, things like `file://` for a standard SQLite file and `libsql://` for libsql.

Perhaps something like this:

```yaml
sqlite:
  dsn: 'libsql://my-libsql-url'
```

And perhaps if it does not start contain a scheme or the scheme is `file://` we could default to SQLite using the current modernc driver? The following being equivalent:

```yaml
sqlite:
  dsn: "schema.sqlite"
```

```yaml
sqlite:
  dsn: "file://schema.sqlite"
```

**Comment by stephenafamo** [comment]
This would be a good way to do it.  
If the dsn starts with `libsql://`, then use the libsql driver.

The other concerns would be

1. Make sure to use the non-CGO driver. 
2. How best to test this libsql driver? Or do we only test SQLite and assume that libsql will remain 100% compatible?

**Comment by stephenafamo** [comment]
Another thing I have just noticed.

The current sqlite driver supports [attaching databases](https://www.sqlite.org/lang_attach.html), and the existing test tries to do this.  
However, attaching databases in libsql works slightly differently and has to be a remote database, attached within a transaction <https://docs.turso.tech/features/attach-database>. This may cause an error when trying to test.

**Comment by plunkettscott** [comment]
Yeah, we don't currently use ATTACH, but that could be a use case to watch out for. Because of the way you've structured config for ATTACH in the generator, I don't think that will be much of a hurdle. You can also operate a full sqld server (the actual server component of LibSQL) locally or in CI using Docker. I would imagine this would be as easy to spin up and seed with data for testing as PostgreSQL and other database engines currently being tested.

Here's a minimal Docker compose to spin u a sqld instance for your reference:

```yaml
services:
  sqld:
    image: ghcr.io/tursodatabase/libsql-server:v0.24.27
    environment:
      - SQLD_NODE=primary
      - SQLD_ADMIN_LISTEN_ADDR=0.0.0.0:8082
    ports:
      - "8080:8080"
      - "8082:8082"
    command:
      - "/bin/sqld"
      - --enable-namespaces
      - --disable-default-namespace
      - --enable-http-console
```

From there, you can create databases for testing like so:

```
POST http://localhost:8082/v1/namespaces/schema/create

Accept: application/json
Content-Type: application/json

{
  "namespace": "a-namespace"
}
```

The connection string for that database would roughly be:

```
libsql://a-namespace.localhost:8080
```

Creating a second database, you could attach the above database using ATTACH.

`ATTACH a-namespace`

Does that help understand how this could be tested? Instead of providing a file path, you provide the other database ID to attach to. You'd need to have both databases created before attempting to test and any necessary schemas applied as well.

Let me know if anything isn't clear, I'm happy to provide more context.

**Comment by plunkettscott** [comment]
> This would be a good way to do it. If the dsn starts with `libsql://`, then use the libsql driver.
> 
> The other concerns would be
> 
>     1. Make sure to use the non-CGO driver.
> 
>     2. How best to test this libsql driver? Or do we only test SQLite and assume that libsql will remain 100% compatible?

1. Yeah, no reason to use the CGO driver. The CGO version is really only necessary if you're testing embedded replicas, but that's a runtime concern and not a schema concern as the schema is the same in that case.

2. I answered how to test this previously, but aside from how ATTACH works behind the scenes and that one is remote and one is a file, there really are not currently differences in the overall schema or dialect. It would likely be best to test with `sqld` though to ensure the driver works as expected in the generator.

**Comment by stephenafamo** [comment]
This is great... I think this should be fairly straightforward then.

1. Use the (non-CGO) libsql driver if the dsn begins with `libsql://`
2. Test the libsql driver in CI with `sqld`.

**Comment by mbezhanov** [comment]
I spent some time looking into this, but I think some things require further clarification before this can be properly implemented (needless to say, I am willing to implement this :slightly_smiling_face:).

From [this announcement](https://turso.tech/blog/upcoming-changes-to-the-turso-platform-and-roadmap#refocusing-database-capabilities) it seems that Turso plans to drop (or perhaps rework?) the `ATTACH` functionality, so it's uncertain how this decision will impact the usage of `ATTACH` statements in the future. Their documentation already indicates that `ATTACH` is deprecated.

Nevertheless, `ATTACH` operations still function correctly on local `sqld` instances started with the `--enable-namespaces` flag:

```bash
docker run -p 8080:8080 -p 8000:8000 --rm \
    ghcr.io/tursodatabase/libsql-server sqld \
    --http-listen-addr=0.0.0.0:8080 \
    --admin-listen-addr=0.0.0.0:8000 \
    --enable-namespaces
```

Since the database that we attach in tests is called `one`, we can create a new namespace with the same name by running the following command:

```bash
curl  -X POST http://localhost:8000/v1/namespaces/one/create \
  -H "Content-Type: application/json" \
  --data '{}' 
```

We also need to make sure that `allow_attach` is set to `true` for this namespace (it defaults to `false`):

```bash
curl -X POST http://localhost:8000/v1/namespaces/one/config \
  -H "Content-Type: application/json" \
  --data '{
    "block_reads": false,
    "block_writes": false,
    "block_reason": null,
    "max_db_size": "1000.0 PB",
    "heartbeat_url": null,
    "jwt_key": null,
    "allow_attach": true,
    "txn_timeout_s": null,
    "durability_mode": "relaxed"
}'
```

Otherwise, we'd get the following error when trying to execute `ATTACH`:

> Authorization forbidden: Namespace `one` doesn't allow attach

As you noted earlier, [this document](https://docs.turso.tech/features/attach-database#things-to-know) uncovers two crucial limitations and considerations regarding the usage of `ATTACH` statements:
- `ATTACH` statements can only be used within database transactions.
- The attached databases are read-only.

When connecting, the `libsql` driver can establish either an HTTP connection or a websocket connection (if the DSN starts with `libsql://` it will default to HTTPs).

However, if we connect over HTTP(s) we'd get the following error trying to start a new transaction:

> interactive transaction not allowed in HTTP queries

Put differently, transactions aren't supported over HTTP, and an `ATTACH` statement sent over HTTP has no real effect on the subsequent queries, so to get the tests working, we have to connect over a websocket (e.g., `ws://localhost:8000`).

Attachment then works, but since the database is attached in read-only mode, migrations start failing with the following error:

> SQLite error: attempt to write a readonly database

To work around this issue, we can split the `schema.sql` file into two smaller files (one for each database) and modify the test code to migrate the two databases independently. 

However, if we plan to do this over a websocket connection, we encounter another issue (described in https://github.com/tursodatabase/go-libsql/issues/22) - the `Exec` method doesn't support multiple statements for websocket connections. So, we either have to adjust our migration helper to parse the contents of each `schema.sql` file into separate `Exec` statements, OR use HTTP connections for migrating and a WS connection for evaluating.

Other than that, the changes to the actual `sqlite.go` driver are minimal. The only thing I haven't yet figured out is how to properly test the regex support, as we can't use `sqlite.RegisterScalarFunction` with the `libsql` driver, but I was able to make all other tests work.

How do you feel about all of this @stephenafamo?

**Comment by mbezhanov** [comment]
Thanks for the idea, that worked 🚀 

I'll start preparing a PR.
